Pod html refactoring 5 of 5

jkeenan opened the pull request:
This if the fifth (and last) in a series of five pull requests submitted with respect to https://github.com/Perl/perl5/issues/18894.

In this p.r., we at long last complete the conversion of the data structures holding information needed to generate HTML files from documents containing POD.  As noted in https://github.com/Perl/perl5/issues/18894, when we started (in the version which shipped with perl-5.34.0), "the code in `ext/Pod-Html/lib/Pod/Html.pm` [was] not fully encapsulated. State [was] held in a total of 19 lexical variables declared outside the scope of the main function, `pod2html()`."  Now that state is held in a `Pod::Html` object and the exportable function `pod2html()` is simply a wrapper around method calls on that object.  The length of `pod2html()` is reduced from 182 lines to 42 lines, which should make it easier to understand what it's doing.

Once this p.r. is merged to blead, we can evaluate patches for all the small points which we're raised in review of this series of pull requests.

jkeenan committed:
Start to make Pod::Html object-oriented.

For the time being, at least, we'll have pod2html() call a constructor.
All subs up through refine_globals have been turned into method calls.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
File::Spec::Unix is not needed within package Pod::Html

It's only needed within Pod::Simple::XHTML::LocalPodLinks, so let's
remove it from the first package.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Make get_cache(), cache_key() and load_cache() use object

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Have generate_cache() use object

Still have to deal with %Pages being outside the object.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Turn 3 more internal subroutines into method calls

identify_input(); parse_input_for_podtree(); set_Title_from_podtree();

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Convert 3 more subroutines to method calls

refine_parser(); feed_tree_to_parser(); write_file()

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Replace _save_pages() with _transform()

_save_pages() is a Pod::Simple::Search method which takes two specific
arguments.  From the point of view of making Pod-Html OO, it was
problematic because (a) it used a variable ($Podroot) defined outside of
its scope and (b) it assigned to a variable (%Pages) defined outside of
its scope.  It was therefore resistant to encapsulation.

Experimentation showed that if we use the return value of
Pod::Simple::Search::survey(), we could parse that hashref using what
was the guts of _save_pages() and assign explicitly to %Pages.

TODO:  Move %Pages within the Pod::Html object.  Handle $object
properly.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

Remove merge conflict marks, commented-out code, as noted by rjbs in
https://github.com/Perl/perl5/pull/19050#discussion_r693559914.

jkeenan committed:
Move $output into the object

Correct dropped variable declaration.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce $self->{Pages}

htmldir3.t: 3rd test unit is now expected to fail.  So now the test file
demonstrates that the unit fails whether it's placed before the second
unit or after.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Increment $VERSION to 1.33 in all .pm files

jkeenan committed:
Pod::Html: assign directly to array ref

This restores the thrust of c0ac28175b, which got zapped when resolving
merge conflicts.

Grinnz commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -558,36 +554,36 @@ sub set_Title_from_podtree {
     }
 
-    $globals->{Title} //= "";
-    return html_escape($globals->{Title});
```
This HTML escaping of the Title appears to get lost.

jkeenan replied:
> This HTML escaping of the Title appears to get lost.

Accepted; handled in 74fb96b24dc12c876b3f2b395418843d6d2d3a47

Grinnz commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -239,237 +238,234 @@ This program is distributed under the Artistic License.
+        $self->{Htmlfileurl} = unixify($self->{Htmlfile});
     }
-    return $globals;
+    return { %{$self} };
```
This seems unnecessary since the return value isn't used anymore. Could just `return $self` like some of the others

jkeenan replied:
Accepted; handled in 34468cc07cbee4d4a1de7ff26d3663a2dabd0fa7

Grinnz commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -239,237 +238,234 @@ This program is distributed under the Artistic License.
     my ($name2path, $path2name) = 
-        Pod::Simple::Search->new->inc(0)->verbose($globals->{Verbose})->laborious(1)
-        ->callback(\&_save_page)->recurse($globals->{Recurse})->survey(@{$globals->{Podpath}});
-    #print STDERR Data::Dumper::Dumper($name2path, $path2name) if ($globals->{Verbose});
```
This comment gets removed but seems to be the only reason Data::Dumper is required when the verbose option is passed. Maybe that could be removed as well.

Grinnz commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -239,237 +238,234 @@ This program is distributed under the Artistic License.
-    my $globals = shift;
-    require Data::Dumper if $globals->{verbose};
+    my $self = shift;
+    require Data::Dumper if $self->{verbose};
```
Data::Dumper required here.

jkeenan replied:
Accepted; handled in 1c6c07161d40fc4b5a3762c4a71cbfc8e62e881d

Grinnz commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -239,237 +238,234 @@ This program is distributed under the Artistic License.
     local $_;
 
-    my $globals = init_globals();
+    my $self = Pod::Html->new();
```
Could be written `__PACKAGE__->new`

jkeenan replied:
The way it's written is more self-documenting, so I think we should leave it this way for now.

jkeenan committed:
Pod::Html: remove debugging artifact

As noted by @Grinnz in
https://github.com/Perl/perl5/pull/19050#discussion_r688969538.

jkeenan committed:
Pod::Html: remove debugging artifact

As noted by @Grinnz in https://github.com/Perl/perl5/pull/19050#discussion_r688969025.

jkeenan committed:
Pod::Html: correct line dropped in 13f1edeb52

As noted by @Grinnz in
https://github.com/Perl/perl5/pull/19050#discussion_r688968921.

rjbs commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -384,36 +384,36 @@ sub refine_globals {
-        Pod::Simple::Search->new->inc(0)->verbose($globals->{Verbose})->laborious(1)
-        ->callback(\&_save_page)->recurse($globals->{Recurse})->survey(@{$globals->{Podpath}});
-    #print STDERR Data::Dumper::Dumper($name2path, $path2name) if ($globals->{Verbose});
+        Pod::Simple::Search->new->inc(0)->verbose($self->{Verbose})->laborious(1)
```
Thanks to serendipity for letting me see that there's a boolean property called `laborious` somewhere.  I got a chuckle.

jkeenan replied:
> Thanks to serendipity for letting me see that there's a boolean property called `laborious` somewhere. I got a chuckle.

The Pod::Simple stuff was the most laborious part of this whole project.

rjbs commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -239,237 +238,233 @@ This program is distributed under the Artistic License.
-    my $_updirs_only = ($globals->{Podroot} =~ /\.\./) && !($globals->{Podroot} =~ /[^\.\\\/]/);
-    foreach my $key (keys %{$Pagesref}) {
+    print $cache join(":", @{$self->{Podpath}}) . "\n$self->{Podroot}\n";
+    my $_updirs_only = ($self->{Podroot} =~ /\.\./) && !($self->{Podroot} =~ /[^\.\\\/]/);
```
```suggestion
    my $_updirs_only = $self->{Podroot} =~ /\.\./ && $self->{Podroot} !~ /[^\.\\\/]/;
```

These kinds of lines can be tricky, so maybe you will not agree that this is a little better.  But if you agree, please feel free to accept this suggestion.

jkeenan replied:
> These kinds of lines can be tricky, so maybe you will not agree that this is a little better. But if you agree, please feel free to accept this suggestion.

Again, following the philosophy of "minimum number of changes needed to get us from Pod-Html as it appears in 5.34.0 to an OO-ified module," I recommend deferring this change until after this p.r. is in blead.  Added to TODO list.

rjbs commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -563,12 +562,12 @@ sub set_Title_from_podtree {
                 (ref($podtree->[4]) eq "ARRAY" &&
                 $podtree->[4]->[0] eq "head1"))) {
-            $globals->{Title} = join("", @{$podtree->[3]}[2..$#{$podtree->[3]}]);
+            $self->{Title} = join("", @{$podtree->[3]}[2..$#{$podtree->[3]}]);
```
prime candidate to be postfix

```
join("", $podtree->[3]->@[ 2 .. $podtree->[3]->$#* ]);
```

Or, what I'd actually write:
```
my @want_indices = (2 .. $podtree->[3]->$#*);
… = join("", $podtree->[3]->@[ $want_indices ]);
```

I have refrained from a lot of "but postfix deref!" but in this case it seems to be to be a significant improvement — but only in clarify, not correctness.

jkeenan replied:
> prime candidate to be postfix
> 
> ```
> join("", $podtree->[3]->@[ 2 .. $podtree->[3]->$#* ]);
> ```
> 
> Or, what I'd actually write:
> 
> ```
> my @want_indices = (2 .. $podtree->[3]->$#*);
> … = join("", $podtree->[3]->@[ $want_indices ]);
> ```
> 
> I have refrained from a lot of "but postfix deref!" but in this case it seems to be to be a significant improvement — but only in clarify, not correctness.

In keeping with the approach I've been taking throughout this project, I'd like to defer introducing postfix deref (where appropriate) until after we've gotten the OO-ified Pod::Html in blead.  That way, if using that proves at all problematic, we will have isolated the problems.  I will add this to the TODO.

rjbs commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -649,6 +646,16 @@ sub write_file {
+#    local $Data::Dumper::Sortkeys = 1;
+#    return (Dumper($ha) eq Dumper($hb)) ? 1 : 0;
+#}
+>>>>>>> b4b3ed877e (Replace _save_pages() with _transform())
```
It appears you have committed a conflicted hunk.  **Surely** this is a mistake!

jkeenan replied:
> It appears you have committed a conflicted hunk. **Surely** this is a mistake!

This is puzzling.  I don't see this in the code I pushed to github.  Github is saying that the lines in question are "Outdated".

I just rebased the branch on blead.  Can you double-check?

jkeenan replied:
@rjbs, I ack-ed the branch for merge-conflict marks and found none.  I also built `perl` at each commit in the branch and ran the tests in `ext/Pod-Html/t/*.t`.  I got no failures.  So unless I hear from you otherwise, I'm going to assume that what you observed poses no barrier to merging this ticket into blead.

rjbs replied:
Your branch includes the commit 47a5a69b5a46d5e567358f2e10594fcca0698b9a which commits merge markers.  A later commit removes them.

jkeenan replied:
> Your branch includes the commit [47a5a69](https://github.com/Perl/perl5/commit/47a5a69b5a46d5e567358f2e10594fcca0698b9a) which commits merge markers. A later commit removes them.

Y'know, I took pains in this project to have small(-ish) commits within each pull request -- but I honestly didn't expect anyone to review on a commit-by-commit basis.  This problem should be repaired now; please double-check.  Thanks for the detailed code review.

rjbs replied:
Looking good!

Reviewing on a "whole diff" versus "commit by commit" basis can be a tough choice.  We were talking about it at $dayjob just last week.  I like commit-by-commit for bigger branches that are altering existing code, *generally*, and whole-diff for new code or a couple small commits.  It's never clear, though.  I wrote a few comments while I was reviewing your work's early commits that I had to then go back and delete in later commits when I saw that my concern was answered by for the end of the commit.

Trade-offs all the way down…

Thanks for your work on this!

rjbs reviewed (approved):
I made some small suggestions.  If it was me, I would amend the history to remove the merge conflict markers from ever appearing.  They don't demonstrate the development of the code itself, and only show that the programmer made a mistake while building this branch.  Best to drop them, in my humble opinion.

My other small comments not withstanding, this looks great.  Seeing all the `$globals` go away was nice, a real payoff.

jkeenan committed:
Pod-Html: remove commented-out code

Make use of Exporter consistent.
